File: src/main/java/br/com/blz/testjava/products/controller/ProductController.java
package br.com.blz.testjava.products.controller;

import br.com.blz.testjava.products.model.Product;
import br.com.blz.testjava.service.ProductService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;

/**
 * @author pzatta
 */

@RestController
@RequestMapping(path = "/products")
public class ProductController {

    @Autowired
    protected ProductService service;

    @GetMapping("/{sku}")
    @ResponseStatus(HttpStatus.OK)
    public Product get(@PathVariable("sku") long sku) {
        return service.findBySku(sku);
    }

    @GetMapping
    @ResponseStatus(HttpStatus.OK)
    public Iterable<Product> get() {
        return service.findAll();
    }

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    public Product post(@RequestBody Product product) {
        return service.create(product);
    }

    @PutMapping("/{sku}")
    @ResponseStatus(HttpStatus.OK)
    public Product put(@PathVariable("sku") long sku, @RequestBody Product product) {
        return service.update(product);
    }

    @DeleteMapping("/{sku}")
    @ResponseStatus(HttpStatus.OK)
    public void detele(@PathVariable("sku") long sku) {
        service.remove(sku);
    }
}

File: src/main/kotlin/com/example/testkotlin/controller/ProductController.kt
package com.example.testkotlin.controller

import com.example.testkotlin.exception.ProductAlreadyExistsException
import com.example.testkotlin.model.Product
import com.example.testkotlin.service.ProductService
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("products")
class ProductController (val productService: ProductService){

    @GetMapping
    fun findAll(): ResponseEntity<List<Product>> {

        val products = productService.findAll()
        if (products.isNotEmpty()) {
            return ResponseEntity<List<Product>>(products, HttpStatus.OK)
        }

        return ResponseEntity<List<Product>>(HttpStatus.NO_CONTENT)

    }

    @GetMapping("/{sku}")
    fun findBySku(@PathVariable sku: Int): ResponseEntity<Product> {

        val product = productService.findBySku(sku)
        if (product != null) {
            return ResponseEntity<Product>(product, HttpStatus.OK)
        }

        return ResponseEntity<Product>(HttpStatus.NOT_FOUND)
    }

    @PostMapping
    fun post(@RequestBody product: Product): ResponseEntity<Product> {

        if (productService.findBySku(product.sku) != null) {
            throw ProductAlreadyExistsException("Product ${product.sku} already exist")
        }
        productService.save(product)
        return ResponseEntity<Product>(product, HttpStatus.CREATED)

    }

    @PutMapping
    fun put(@RequestBody product: Product): ResponseEntity<Product> {

        if (productService.update(product)) {
            return ResponseEntity<Product>(product, HttpStatus.OK)
        }
        return ResponseEntity<Product>(HttpStatus.NOT_FOUND)

    }

    @DeleteMapping("/{sku}")
    fun delete(@PathVariable sku: Int): ResponseEntity<Boolean> {

        if (productService.delete(sku)) {
            return ResponseEntity<Boolean>(true, HttpStatus.OK)
        }
        return ResponseEntity<Boolean>(false, HttpStatus.NOT_FOUND)

    }
}
